op-node/eth/output.go

```diff
@@ -12,3 +15,70 @@ type OutputResponse struct {
 	StateRoot             common.Hash `json:"stateRoot"`
 	Status                *SyncStatus `json:"syncStatus"`
 }
+
+var (
+	ErrInvalidOutput        = errors.New("invalid output")
+	ErrInvalidOutputVersion = errors.New("invalid output version")
+
+	OutputVersionV0 = Bytes32{}
+)
+
+type Output interface {
+	// Version returns the version of the L2 output
+	Version() Bytes32
+
+	// Marshal a L2 output into a byte slice for hashing
+	Marshal() []byte
+}
+
+type OutputV0 struct {
+	StateRoot                Bytes32
+	MessagePasserStorageRoot Bytes32
+	BlockHash                common.Hash
+}
+
+func (o *OutputV0) Version() Bytes32 {
+	return OutputVersionV0
+}
+
+func (o *OutputV0) Marshal() []byte {
+	var buf [128]byte
+	version := o.Version()
+	copy(buf[:32], version[:])
+	copy(buf[32:], o.StateRoot[:])
+	copy(buf[64:], o.MessagePasserStorageRoot[:])
+	copy(buf[96:], o.BlockHash[:])
+	return buf[:]
+}
+
+// OutputRoot returns the keccak256 hash of the marshaled L2 output
+func OutputRoot(output Output) Bytes32 {
+	marshaled := output.Marshal()
+	return Bytes32(crypto.Keccak256Hash(marshaled))
+}
+
+func UnmarshalOutput(data []byte) (Output, error) {
+	if len(data) < 32 {
+		return nil, ErrInvalidOutput
+	}
+	var ver Bytes32
+	copy(ver[:], data[:32])
+	switch ver {
+	case OutputVersionV0:
+		return unmarshalOutputV0(data)
+	default:
+		return nil, ErrInvalidOutputVersion
+	}
+}
+
+func unmarshalOutputV0(data []byte) (*OutputV0, error) {
+	if len(data) < 128 {
```

Comment:
```suggestion
	if len(data) != 128 {
```
Shouldn't we be more strict about the data size?

Reply:
I wanted to accommodate usecases where the caller has a buffer containing both OutputRootV0 and other data. But this is premature. Will fix.